### PR TITLE
stmhal: implement HID OUT endpoint (USB receive)

### PR DESCRIPTION
**Pull request by prusnak**
This changeset implements HID OUT endpoint, which allows receiving data on the uPy side of things. 

Original commits from https://github.com/philandstuff/micropython/tree/usb-hid-out-endpoint, added few more fixes by me.

### CODE REVIEW

**Comment by prusnak** [comment]
Implements #1217

**Comment by prusnak** [comment]
To test

Put this into `boot.py`:
``` python
import pyb
pyb.usb_mode('VCP+HID')
```

Run this in REPL:

``` python
import pyb
h = pyb.USB_HID()
h.recv(64)
````

Then on PC side do the following:

``` python
import hid
h = hid.device()
h.open(0xf055, 0x9801)
h.write([0] + range(64))
```

recv function should return bytes.

File: stmhal/usbd_hid_interface.c [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        if (query_irq() == IRQ_STATE_DISABLED) {
+            // IRQs disabled so buffer will never be filled; return immediately
+            return 0;
+        }
+        __WFI(); // enter sleep mode, waiting for interrupt
+    }
+
+    // Copy bytes from device to user buffer
+    memcpy(buf, buffer[current_read_buffer], last_read_len);
```
Comment (dpgeorge) [inline stmhal/usbd_hid_interface.c]: The len argument to this function is not used at all, in particular this line could overrun buf if buf is too short.

File: stmhal/usbd_hid_interface.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/*
```
Comment (dpgeorge) [inline stmhal/usbd_hid_interface.h]: This is such a simple file.  The original copyright is not needed.

File: stmhal/usbdev/class/src/usbd_cdc_msc_hid.c [hunk 3]
```diff
@@ -931,8 +967,10 @@ static uint8_t USBD_CDC_MSC_HID_DataOut(USBD_HandleTypeDef *pdev, uint8_t epnum)
     } else if ((usbd_mode & USBD_MODE_MSC) && epnum == (MSC_OUT_EP & 0x7f)) {
         MSC_BOT_DataOut(pdev, epnum);
         return USBD_OK;
+    } else if ((usbd_mode & USBD_MODE_HID) && epnum == (hid_out_ep & 0x7f)) {
+      HID_ClassData.RxLength = USBD_LL_GetRxDataSize(pdev, epnum);
+      HID_fops->Receive(HID_ClassData.RxBuffer, HID_ClassData.RxLength);
     }
-
```
Comment (dpgeorge) [inline stmhal/usbdev/class/src/usbd_cdc_msc_hid.c]: Please keep this blank line.

File: stmhal/usbd_hid_interface.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
+/*
+ * TODO: this header is even more lies now
```
Comment (dpgeorge) [inline stmhal/usbd_hid_interface.c]: Where did the bulk of this file come from, an ST example?

**Comment by prusnak** [comment]
Pushed a new commit fixing the issues.

**Comment by dpgeorge** [comment]
I tested it and there was a bug: needed to adjust size of MAX_TEMPLATE_CONFIG_DESC_SIZE from 100 to 107.

Otherwise it works well.  Thanks for the submission, it's a very cool feature to have!

During the merge I squashed some commits in various ways to try and make the history a bit cleaner.  See 99d62c4def5f62af66cf66b3e620de21f017db29 for last of the 4 commits.

**Comment by prusnak** [comment]
Thanks!

**Pull request closed without merging**